File: Day278/findLargestValueinEachTreeLevel.java
import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> answer;
    public List<Integer> largestValues(TreeNode root) {
        answer = new ArrayList<>();
        dfs(root, -1);
        return answer;
    }
    public void dfs(TreeNode root, int level) {
        if (root == null) {
            return;
        }
        level++;
        if (answer.size() == level) {
            answer.add(root.val); 
        } else {
            answer.set(level, Math.max(root.val, answer.get(level)));
        }
        dfs(root.left, level);
        dfs(root.right, level);
    }
}
 public List<Integer> largestValues(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        List<Integer> answer = new ArrayList<>();
        if (root == null) {
            return answer;
        }
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            int max  =Integer.MIN_VALUE;
            List<Integer> t = new ArrayList<>();
            for (int i=0; i<size; i++) {
                TreeNode temp = queue.poll();
                if (temp != null) {
                    max = Math.max(max, temp.val);
                    if (temp.left != null) {
                        queue.offer(temp.left);
                    }
                    if (temp.right != null) {
                        queue.offer(temp.right);
                    }
                }
            }
            answer.add(max);
        }
        return answer;
    }
}

File: Day278/findLargestValueinEachTreeLevel.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func dfs(root *TreeNode, answer *[]int, level int) {
    if root == nil {
        return
    }
    level++
    if (len(*answer) == level) {
        *answer = append(*answer, root.Val)
    } else {
        (*answer)[level] = int(math.Max(float64((*answer)[level]), float64(root.Val)))
    }
    dfs(root.Left, answer, level)
    dfs(root.Right, answer, level)
}

func largestValues(root *TreeNode) []int {
    answer := make([]int, 0)
    dfs(root, &answer, -1)
    return answer
}